### PR TITLE
mobile: add logging for exception handling

### PR DESCRIPTION
**Pull request by Augustyniak**
Commit Message:
Additional Description: WIP - for demonstration purposes only.
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: mobile/library/common/jni/jni_utility.cc [hunk 1]
```diff
@@ -69,6 +69,7 @@ void jni_delete_const_global_ref(const void* context) {
 bool clear_pending_exceptions(JNIEnv* env) {
   if (env->ExceptionCheck() == JNI_TRUE) {
     env->ExceptionClear();
+    ENVOY_LOG_EVENT_TO_LOGGER(GET_MISC_LOGGER(), info, "jni_exception", "THIS CRASHES");
```
Comment (Augustyniak) [inline mobile/library/common/jni/jni_utility.cc:72]: @alyssawilk Do you have any insight as to whether this line of code would crash here? Notice that I added another `ENVOY_LOG_EVENT_TO_LOGGER` logger in this PR and it does _not_ crash.

These are the logs that I am seeing when our CI job (ran on Linux) fails. The logs mentioned the deadlock but the root cause for the issue is a crash that eventually leads to a deadlock:
```
[2022-12-12 15:17:22.142][40][info][misc] [library/common/extensions/filters/http/platform_bridge/filter.cc:283] THIS WORKS - DOES NOT CRASH
{name=event_log, message=THIS WORKS - DOES NOT CRASH, log_name=before_on_headers}[2022-12-12 15:17:22.143][40][info][misc] [library/common/jni/jni_utility.cc:72] THIS CRASHES
[mutex.cc : 1368] RAW: Potential Mutex deadlock: 
	@ 0x7f16c787fdbf absl::DebugOnlyDeadlockCheck()
	@ 0x7f16c788071f absl::Mutex::ReaderLock()
	@ 0x7f16c53e62c4 absl::ReaderMutexLock::ReaderMutexLock()
	@ 0x7f16c71864a4 Envoy::Logger::DelegatingLogSink::log()
	@ 0x7f16c5381ce2 spdlog::logger::sink_it_()
	@ 0x7f16c53e50ef spdlog::logger::log_it_()
	@ 0x7f16c548c0ca spdlog::logger::log_<>()
	@ 0x7f16c548bdef spdlog::logger::log<>()
	@ 0x7f16c5487d8a Envoy::StrippedMainBase::StrippedMainBase()::$_0::operator()()
	@ 0x7f16c5487a99 Envoy::StrippedMainBase::StrippedMainBase()::$_0::__invoke()
	@ 0x7f16fc47cb50 operator new()

[mutex.cc : 1380] RAW: Acquiring absl::Mutex 0x7f16c8007e88 while holding  0x7f16c8007e88; a cycle in the historical lock ordering graph has been observed
[mutex.cc : 1381] RAW: Cycle: 
[mutex.cc : 1395] RAW: mutex@0x7f16c8007e88 stack: 
	@ 0x7f16c787fdbf absl::DebugOnlyDeadlockCheck()
	@ 0x7f16c788071f absl::Mutex::ReaderLock()
	@ 0x7f16c53e62c4 absl::ReaderMutexLock::ReaderMutexLock()
	@ 0x7f16c718b131 Envoy::Logger::DelegatingLogSink::delegate()
	@ 0x7f16c7185e98 Envoy::Logger::SinkDelegate::setDelegate()
	@ 0x7f16c7186036 Envoy::Logger::StderrSinkDelegate::StderrSinkDelegate()
	@ 0x7f16c718b4ce std::make_unique<>()
	@ 0x7f16c718692a Envoy::Logger::DelegatingLogSink::init()
	@ 0x7f16c53e33c8 Envoy::Logger::Registry::getSink()
	@ 0x7f16c5449a40 Envoy::Engine::main()
	@ 0x7f16c547781e std::__invoke_impl<>()
	@ 0x7f16c54775e1 std::__invoke<>()
	@ 0x7f16c547755d std::thread::_Invoker<>::_M_invoke<>()
	@ 0x7f16c54774c5 std::thread::_Invoker<>::operator()()
	@ 0x7f16c5476e39 std::thread::_State_impl<>::_M_run()
	@ 0x7f16fc4a8de4 

[mutex.cc : 1400] RAW: dying due to potential deadlock
[mutex.cc : 1368] RAW: Potential Mutex deadlock: 
	@ 0x7f16c787fdbf absl::DebugOnlyDeadlockCheck()
	@ 0x7f16c788071f absl::Mutex::ReaderLock()
	@ 0x7f16c53e62c4 absl::ReaderMutexLock::ReaderMutexLock()
	@ 0x7f16c71864a4 Envoy::Logger::DelegatingLogSink::log()
	@ 0x7f16c5381ce2 spdlog::logger::sink_it_()
	@ 0x7f16c53e50ef spdlog::logger::log_it_()
	@ 0x7f16c6d578e5 spdlog::logger::log_<>()
	@ 0x7f16c6d5761a spdlog::logger::log<>()
	@ 0x7f16c6d57473 Envoy::BackwardsTrace::logFault()
	@ 0x7f16c6d53997 Envoy::SignalAction::sigHandler()
	@ 0x7f171cf38420 __restore_rt

[mutex.cc : 1380] RAW: Acquiring absl::Mutex 0x7f16c8007e88 while holding  0x7f16c8007e88; a cycle in the historical lock ordering graph has been observed
[mutex.cc : 1381] RAW: Cycle: 
[mutex.cc : 1395] RAW: mutex@0x7f16c8007e88 stack: 
	@ 0x7f16c787fdbf absl::DebugOnlyDeadlockCheck()
	@ 0x7f16c788071f absl::Mutex::ReaderLock()
	@ 0x7f16c53e62c4 absl::ReaderMutexLock::ReaderMutexLock()
	@ 0x7f16c718b131 Envoy::Logger::DelegatingLogSink::delegate()
	@ 0x7f16c7185e98 Envoy::Logger::SinkDelegate::setDelegate()
	@ 0x7f16c7186036 Envoy::Logger::StderrSinkDelegate::StderrSinkDelegate()
	@ 0x7f16c718b4ce std::make_unique<>()
	@ 0x7f16c718692a Envoy::Logger::DelegatingLogSink::init()
	@ 0x7f16c53e33c8 Envoy::Logger::Registry::getSink()
	@ 0x7f16c5449a40 Envoy::Engine::main()
	@ 0x7f16c547781e std::__invoke_impl<>()
	@ 0x7f16c54775e1 std::__invoke<>()
	@ 0x7f16c547755d std::thread::_Invoker<>::_M_invoke<>()
	@ 0x7f16c54774c5 std::thread::_Invoker<>::operator()()
	@ 0x7f16c5476e39 std::thread::_State_impl<>::_M_run()
	@ 0x7f16fc4a8de4 

[mutex.cc : 1400] RAW: dying due to potential deadlock
```

Comment (alyssawilk) [reply mobile/library/common/jni/jni_utility.cc:72]: not sure - I haven't seen any issues with logger locks before so I assume it's an E-M specific quirk.
I do notice that all those functions log via jni_log, which I would assume is the correct thing to do here only that seems to be a no-op by default?

Comment (Augustyniak) [reply mobile/library/common/jni/jni_utility.cc:72]: @danzh2010 this is an issue (deaddlock) that I ran into the last time I tried to log from within the JNI layer to the platform layer. It does not seem to be related to an uncleared pending exception.

Comment (Augustyniak) [reply mobile/library/common/jni/jni_utility.cc:72]: I cannot see this deadlock anymore in https://github.com/envoyproxy/envoy/pull/26133 I think. Going to look at it closer tmrw.

Comment (danzh2010) [reply mobile/library/common/jni/jni_utility.cc:72]: I think it depends on where `clear_pending_exceptions()` gets called. It seems that an exception happened during start up: https://github.com/envoyproxy/envoy/blob/main/mobile/library/common/engine.cc#L71 where a mutex lock is already hold according to the [call stack](https://github.com/envoyproxy/envoy/pull/24488/files#:~:text=sigHandler()%0A%09%40%200x7f171cf38420%20__restore_rt-,%5Bmutex.cc%20%3A%201380%5D%20RAW%3A%20Acquiring%20absl%3A%3AMutex%200x7f16c8007e88%20while%20holding%20%200x7f16c8007e88,_M_run()%0A%09%40%200x7f16fc4a8de4%20%0A%0A%5Bmutex.cc%20%3A%201400%5D%20RAW%3A%20dying%20due%20to%20potential%20deadlock,-Member) above. Can you print out the exception details before calling `ExceptionClear()` here to see where did it throw? 

Anyway very likely exceptions happens after start up won't run into this dead lock.

Comment (Augustyniak) [reply mobile/library/common/jni/jni_utility.cc:72]: @danzh2010 thank you for taking a look - that's helpful. Unfortunately (?) I cannot repro the crash using previously working repro-step anymore so something had to change - it's possible that the ordering of some operations changed or some operation was changed from being sync to async or vice versa or we stopped throwing during engine lunch.

Going to add try to simulate a crash during engine start up to see whether it allows me to repro.

**Pull request closed without merging**